src/Type/Php/ImplodeFunctionReturnTypeExtension.php

```diff
@@ -115,13 +116,21 @@ private function inferConstantType(ConstantArrayType $arrayType, ConstantStringT
 
 			$arrayValues = [];
 			foreach ($valueTypes as $valueType) {
-				if (!$valueType instanceof ConstantScalarType) {
+				$constScalars = $valueType->getConstantScalarValues();
+				if (count($constScalars) === 0) {
 					return null;
 				}
-				$arrayValues[] = $valueType->getValue();
+				$arrayValues[] = $constScalars;
 			}
 
-			$strings[] = new ConstantStringType(implode($separatorType->getValue(), $arrayValues));
+			$combinations = CombinationsHelper::combinations($arrayValues);
+			foreach ($combinations as $combination) {
+				$strings[] = new ConstantStringType(implode($separatorType->getValue(), $combination));
+			}
+		}
+
+		if (count($strings) > InitializerExprTypeResolver::CALCULATE_SCALARS_LIMIT) {
```

Comment:
Ideally you'd count the combinations before generating them, it can explode as well.